src/updater/updater.cpp

```diff
@@ -149,446 +95,568 @@ void CheckWeightsConsistency(
+            storage::io::FileReader profile_properties_file(
+                config.profile_properties_path, storage::io::FileReader::HasNoFingerprint);
+            profile_properties = profile_properties_file.ReadOne<extractor::ProfileProperties>();
+        };
 
-    // Mapped file pointer for turn indices
-    const extractor::lookup::TurnIndexBlock *turn_index_blocks =
-        reinterpret_cast<const extractor::lookup::TurnIndexBlock *>(
-            turn_penalties_index_region.get_address());
-    BOOST_ASSERT(is_aligned<extractor::lookup::TurnIndexBlock>(turn_index_blocks));
+        tbb::parallel_invoke(load_edge_data,
+                             load_segment_data,
+                             load_turn_weight_penalties,
+                             load_turn_duration_penalties,
+                             load_profile_properties);
+    }
 
-    // Mapped file pointers for edge-based graph edges
-    auto edge_based_edge_ptr = reinterpret_cast<const extractor::EdgeBasedEdge *>(
-        reinterpret_cast<char *>(edge_based_graph_region.get_address()) +
-        sizeof(EdgeBasedGraphHeader));
-    BOOST_ASSERT(is_aligned<extractor::EdgeBasedEdge>(edge_based_edge_ptr));
+    tbb::concurrent_vector<GeometryID> updated_segments;
+    if (update_edge_weights)
+    {
+        auto segment_speed_lookup = csv::readSegmentValues(config.segment_speed_lookup_paths);
 
-    auto edge_segment_byte_ptr = reinterpret_cast<const char *>(edge_segment_region.get_address());
+        TIMER_START(segment);
+        updated_segments =
+            updateSegmentData(config, profile_properties, segment_speed_lookup, segment_data);
+        // Now save out the updated compressed geometries
+        extractor::io::write(config.geometry_path, segment_data);
+        TIMER_STOP(segment);
+        util::Log() << "Updating segment data took " << TIMER_MSEC(segment) << "ms.";
+    }
 
-    bool fallback_to_duration = true;
-    for (std::uint64_t edge_index = 0; edge_index < graph_header.number_of_edges; ++edge_index)
+    if (update_turn_penalties)
     {
-        // Make a copy of the data from the memory map
-        extractor::EdgeBasedEdge inbuffer = edge_based_edge_ptr[edge_index];
+        auto turn_penalty_lookup = csv::readTurnValues(config.turn_penalty_lookup_paths);
+        auto updated_turn_penalties = updateTurnPenalties(config,
+                                                          profile_properties,
+                                                          turn_penalty_lookup,
+                                                          turn_weight_penalties,
+                                                          turn_duration_penalties);
+        // we need to re-compute all edges that have updated turn penalties.
+        // this marks it for re-computation
+        std::transform(updated_turn_penalties.begin(),
+                       updated_turn_penalties.end(),
+                       std::back_inserter(updated_segments),
```

Comment:
I would be cautious with back_inserter (and to some extend push_back), if you do a structure preserving transform and call the inserter for every element anyway, the better alternative is to simply resize and assign.

Here's some anecdotal evidence https://github.com/mattgodbolt/zindex/commit/27d229afc5021f208467c3526b5a223ecf44b6e4

(Not only or specifically here, it just caught my eye)